### PR TITLE
Supports appdynamics via user provided service

### PR DESCRIPTION
**Pull request by kpschuck**
## Proposed change

Modifies the `jq` filter to search for the first occurrence of _any_ service who's name contains `appdynamics` to source credentials from.

## Use cases the change solves

This allows a user provided service or brokered service with a name containing `appdynamics` to be used instead of _only_ a brokered service.

* [x] I have viewed signed and have submitted the Contributor License Agreement

* [x] I have made this pull request to the `develop` branch

* [ ] I have added an integration test

I'll take a look at adding an integration test and getting the CLA signed. Just wanted to get it out here for comment in the meantime.

### CODE REVIEW

**Comment by sclevine** [comment]
Hi @kpschuck, thanks for contributing. We're definitely open to a PR that adds the functionality you describe.

**Comment by sclevine** [comment]
@kpschuck do you still plan to sign the CLA?

**Comment by kpschuck** [comment]
@sclevine -- just emailed it

**Comment by paulbeattie** [comment]
Nice to see some work being done on this, this introduces the bug where if you had a service say a proxy service which was implemented to allow communication to AppDynamics SaaS controller and you named it `appdynamics-proxy` this would then catch the proxy service and try to extract auth information for AppD from there. Surely the correct behaviour for this is to check for the service type rather than just the name?

**Comment by sclevine** [comment]
@paulbeattie I don't think the service type is exposed via a user-provided service. It does seem like we should be more careful not to match other UPSs accidentally.

What if we only matched on the string `appdynamics-ups[-]?(.*)`?

**Comment by nkatre** [comment]
Agree with the changes by @kpschuck . I think with this change, the only requirement is that the name "appdynamics" should be present in the cups service name. I dont see any other proper way of identifying the service correctly. We should get this merged asap

**Comment by sclevine** [comment]
We decided to proceed by matching names that start with `appdynamics-ups` instead. Otherwise, we risk binding to non-appdynamics services. Is there any issue with requiring the extra `-ups` part?

**Comment by nkatre** [comment]
In which cases would this happen where searching for `appdynamics` risk binding with non-appdynamics service ?

**Comment by nkatre** [comment]
I think just appdynamics is much cleaner than the extra ups part, Moreover the services created by appdynamics tile has the keywords as `appdynamics` and these directly added services might be broken since both uses the same script

**Comment by sclevine** [comment]
The service bindings to service instances created by the appdynamics tile will have a `type` key that will be matched separately.

This change only applies to user-provided services, and it seems risky to blindly match any service with `appdynamics` in the name.

That said, it wasn't clear to me if @paulbeattie's example was a common use case. If we have no reason to suspect that anyone would bind a service with appdynamics in the name, I could be convinced to match on service names that start with `appdynamics` instead.

@paulbeattie, does the knowledge that we can't match on service type change your opinion?

**Comment by kpschuck** [comment]
The intention of this PR was to allow matching on either UPS or brokered services instead of only brokered services. Since the existing behavior does not support the `app-dynamics` substring, is it in scope for this PR to fix that as well? 

You'll notice the example from the PHP buildpack doesn't quite work since it filters on `appdynamics` or `app-dynamics` to detect, then only checks for a marketplace service called `appdynamics` (https://github.com/cloudfoundry/php-buildpack/blob/ecef9d90bf40f598862b4aad19b2eacedf0f4619/extensions/appdynamics/extension.py#L93).

In summary, is the intended behavior to do the following?
  1. Check for the existence of a marketplace service with the substring `app-dynamics` or `appdynamics`
  2. If not found, search UPSs for a service names with the substring `app-dynamics` or `appdynamics`
  3. [Pick the first match](https://github.com/cloudfoundry/php-buildpack/blob/master/extensions/appdynamics/extension.py#L101)

**Comment by nkatre** [comment]
@kpschuck The marketplace service is configured as `appdynamics`. Actually we do not need to check for `app-dynamics`, I don't know why the filter in the place was checking for `app-dynamics`. I think its safe to remove `app-dyanmics` from the fitler

**Comment by sclevine** [comment]
If a marketplace service binding  with type `appdynamics` (exact string) is found, we should use that.

Otherwise, we should pick the first user-provided service binding with the substring `appdynamics` or `app-dynamics` in the binding name. We will keep `app-dynamics` for user-provided services to maintain compatibility with the other buildpacks that accept that substring.

**Comment by sclevine** [comment]
@kpschuck Feel free to update the PR to match those criteria, but also note that the core buildpacks team will adjust the PR accordingly when they reach the corresponding tracker story in the buildpacks backlog.

**Comment by paulbeattie** [comment]
@sclevine @kpschuck sorry this got buried under a pile of other stuff earlier. Great discussion too! Shame I missed out. Totally happy to do as outlined in https://github.com/cloudfoundry/nodejs-buildpack/pull/106#issuecomment-342930227 that should cover most cases 🙌

**Comment by nkatre** [comment]
@sclevine @paulbeattie When are these changes scheduled to be merged for next release ?

**Comment by sclevine** [comment]
The first release that will contain these changes can be followed here:
https://www.pivotaltracker.com/story/show/152816054

**Comment by sclevine** [comment]
@nkatre Any progress on getting us an AppDynamics account to test this integration?

**Comment by nkatre** [comment]
@sclevine I am still working with the team to provide this.I should have an update by the end of the week
